cumulus_etl/loaders/fhir/bulk_export.py

```diff
@@ -236,90 +233,133 @@ async def _delete_export(self, poll_url: str) -> None:
+        Requests a file, while respecting any requests to wait longer and telling the user.
 
         :param path: path to request
         :param headers: headers for request
         :param target_status_code: retries until this status code is returned
         :param method: HTTP method to request
+        :param log_begin: method to call to report that we are about to start requests
+        :param log_request: method to call to report every request attempt
+        :param log_progress: method to call to report a successful request but not yet done
+        :param log_error: method to call to report request failures
+        :param stream: whether to stream the response
+        :param retry_errors: if True, server-side errors will be retried a few times
         :returns: the HTTP response
         """
-        status_box = rich.text.Text()
-        with rich.get_console().status(status_box) as status:
-            while self._total_wait_time < self._TIMEOUT_THRESHOLD:
-                response = await self._client.request(method, path, headers=headers)
-
-                if response.status_code == target_status_code:
-                    if status_box.plain:
-                        status.stop()
-                        print(
-                            f"  Waited for a total of {common.human_time_offset(self._total_wait_time)}"
-                        )
-                    return response
-
-                # 202 == server is still working on it, 429 == server is busy.
-                # In both cases, we wait.
-                if response.status_code in [202, 429]:
-                    if log_progress:
-                        log_progress(response)
-
-                    # Print a message to the user, so they don't see us do nothing for a while
-                    delay = int(response.headers.get("Retry-After", 60))
-                    if response.status_code == 202:
-                        # Some servers can request unreasonably long delays (e.g. I've seen Cerner ask for five hours),
-                        # which is... not helpful for our UX and often way too long for small exports.
-                        # So as long as the server isn't telling us it's overloaded, limit the delay time to 5 minutes.
-                        delay = min(delay, 300)
-                    progress_msg = response.headers.get("X-Progress", "waiting…")
-                    formatted_total = common.human_time_offset(self._total_wait_time)
-                    formatted_delay = common.human_time_offset(delay)
-                    status_box.plain = f"{progress_msg} ({formatted_total} so far, waiting for {formatted_delay} more)"
+        # Set up error handling variables.
+        # These times are extremely generous - partly because we can afford to be
+        # as a long-running async task and partly because EHR servers seem prone to
+        # outages that clear up after a bit.
+        error_retry_minutes = [1, 2, 4, 8]  # and then raise
```

Comment:
is there any value in allowing a user to specify the number of retries?

Reply:
So far, I've just put retrying on the "application" side of the http code - i.e. not directly in FhirClient, but rather in the BulkExporter code.

Partly because I wanted to give myself the freedom to be a bit inflexible with it for now. Like... I dunno, is the user really going to want to specify number of retries? Feels too nitty gritty - if after 15minutes, a server still has (recoverable) problems, I'd rather just hear about it and then add another hard-coded retry.

The other odd thing about this retry code is that it's (intentionally) so slow. Most "exponential backoff" code out there is like, "wait 2 seconds, then 4, then 20 seconds" and I'm out here waiting 8 minutes.

So when I was thinking of building flexibility into FhirClient for number of retries and how long, just to use it in one place with weird values, I figured - just build the inflexible version first.

If we want retrying elsewhere in the ETL later, we can push this down a layer into FhirClient, with more flexibility. Does that sound OK?

Reply:
yeah that's perfectly reasonable